### ISSUE TITLE
`RUF055` safe fix changes behavior for non-string inputs

### ISSUE THREAD

**Issue by ntBre** [astral-sh/ruff#27024]
### Summary

The RUF055 fix is classified as safe, but it changes behavior when a bytes regular expression searches an object that implements the buffer protocol.

[Playground reproduction](https://play.ruff.rs/4b61dff1-5a7e-4e67-a671-42ef5d4a2f96)

```python
import re


assert re.search(b"ab", memoryview(b"abc"))
```

Ruff replaces the assertion with:

```python
assert b"ab" in memoryview(b"abc")
```

The original assertion passes. The replacement fails because `re.search` accepts the `memoryview` as a bytes-like search target, while membership on a `memoryview` does not perform a byte-subsequence search.

I think the rule probably should not apply unless we can confirm that the target is of the proper type, or at least we shouldn't offer a fix. This blocks #26908.

**Comment by soreavis** [comment]
Fair enough on #27233 — I made the scoping call myself instead of asking.

One data point in case it's useful: on five files from a real site-packages, all 9 current RUF055 hits are str patterns, and only one has a target Ruff can infer. So a gate that also covers unknown targets for str patterns drops 8 of the 9.

Would you rather suppress those outright, or go with the no-fix option you suggested — keep the diagnostic, drop the fix? I can build either.
